### PR TITLE
feat(mojaloop/#3904): add position event timeout reserved batch handling

### PR DESCRIPTION
**Pull request by kleyow**
feat(mojaloop/#3904): add position event timeout reserved batch handling

### CODE REVIEW

File: src/domain/position/timeout-reserved.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  const transferStateId = Enum.Transfers.TransferInternalState.EXPIRED_RESERVED
+  // Revert payer's position for the amount of the transfer
+  // NOTE: We are pulling the transferAmount using
+  // const latestTransferInfoByTransferId = await BatchPositionModel.getTransferInfoList(
+  //   trx,
+  //   transferIdList,
+  //   Enum.Accounts.TransferParticipantRoleType.PAYEE_DFSP,
+  //   Enum.Accounts.LedgerEntryType.PRINCIPLE_VALUE
+  // )
+  // Is it safe using this value of the payee even accounting for the negative value?
```
Comment (kleyow) [inline src/domain/position/timeout-reserved.js:146]: :thinking:

File: src/domain/position/timeout-reserved.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    Enum.Kafka.Topics.POSITION,
+    Enum.Events.Event.Action.TIMEOUT_RESERVED,
+    state
+  )
+  const resultMessage = Utility.StreamingProtocol.createMessage(
+    transferId,
+    payeeFsp,
+    payerFsp,
+    metadata,
+    headers, // Unsure what headers need to be passed here
```
Comment (kleyow) [inline src/domain/position/timeout-reserved.js:127]: :thinking: couldn't tell from the non batch code what the headers for this are.

File: src/domain/position/timeout-reserved.js (hunk 1 above)
Comment (kleyow) [reply src/domain/position/timeout-reserved.js:146]: After dicussions this value is safe to use to revert the payer's position.

File: src/domain/position/timeout-reserved.js (hunk 2 above)
Comment (kleyow) [reply src/domain/position/timeout-reserved.js:127]: The ml-api-adapter deals with notifying the payee and payer, so headers don't matter and the adapter will construct the proper headers!
https://github.com/mojaloop/ml-api-adapter/blob/main/src/handlers/notification/index.js#L545

As long as the to and from are the payer and payee, which they have been and the action is timeout-reserved ml adapter will notify both.

File: src/domain/position/binProcessor.js [hunk 3]
```diff
@@ -129,6 +137,24 @@ const processBins = async (bins, trx) => {
...
+
+    // Update accumulated values
+    accumulatedPositionValue = timeoutReservedActionResult.accumulatedPositionValue
+    accumulatedPositionReservedValue = timeoutReservedActionResult.accumulatedPositionReservedValue
+    accumulatedTransferStates = timeoutReservedActionResult.accumulatedTransferStates
+    // Append accumulated arrays
+    accumulatedTransferStateChanges = accumulatedTransferStateChanges.concat(timeoutReservedActionResult.accumulatedTransferStateChanges)
+    accumulatedPositionChanges = accumulatedPositionChanges.concat(timeoutReservedActionResult.accumulatedPositionChanges)
+    notifyMessages = notifyMessages.concat(timeoutReservedActionResult.notifyMessages)
+
```
Comment (vijayg10) [inline src/domain/position/binProcessor.js:157]: I think this one should come after fulfil and before prepare according to this diagram
Screenshot 2024-05-17 at 2 18 16 PM

File: src/domain/position/binProcessor.js [hunk 4]
```diff
@@ -105,8 +106,15 @@ const processBins = async (bins, trx) => {
...
     // We need to remove this once we implement all the actions
-    if (!isSubset([Enum.Events.Event.Action.PREPARE, Enum.Events.Event.Action.FX_PREPARE, Enum.Events.Event.Action.COMMIT, Enum.Events.Event.Action.RESERVE, Enum.Events.Event.Action.FX_RESERVE], actions)) {
-      Logger.isErrorEnabled && Logger.error('Only prepare/fx-prepare/commit actions are allowed in a batch')
+    if (!isSubset([
+      Enum.Events.Event.Action.PREPARE,
+      Enum.Events.Event.Action.FX_PREPARE,
+      Enum.Events.Event.Action.COMMIT,
+      Enum.Events.Event.Action.RESERVE,
+      Enum.Events.Event.Action.FX_RESERVE,
+      Enum.Events.Event.Action.TIMEOUT_RESERVED
```
Comment (vijayg10) [inline src/domain/position/binProcessor.js:115]: Don't we need separate actions like TIMEOUT_RESERVED and TIMEOUT_FX_RESERVED?

File: test/integration-override/handlers/positions/handlerBatch.test.js [hunk 5]
```diff
@@ -66,10 +66,10 @@ const TransferInternalState = Enum.Transfers.TransferInternalState
...
 const TransferEventAction = Enum.Events.Event.Action
 
-const debug = process?.env?.TEST_INT_DEBUG || false
-// const rebalanceDelay = process?.env?.TEST_INT_REBALANCE_DELAY || 10000
-const retryDelay = process?.env?.TEST_INT_RETRY_DELAY || 2
-const retryCount = process?.env?.TEST_INT_RETRY_COUNT || 40
+const debug = process?.env?.test_INT_DEBUG || false
+// const rebalanceDelay = process?.env?.test_INT_REBALANCE_DELAY || 10000
+const retryDelay = process?.env?.test_INT_RETRY_DELAY || 2
+const retryCount = process?.env?.test_INT_RETRY_COUNT || 40
```
Comment (vijayg10) [inline test/integration-override/handlers/positions/handlerBatch.test.js:72]: Why is this change?

File: src/domain/position/binProcessor.js (hunk 3 above)
Comment (kleyow) [reply src/domain/position/binProcessor.js:157]: thanks. done.

File: src/domain/position/binProcessor.js (hunk 4 above)
Comment (kleyow) [reply src/domain/position/binProcessor.js:115]: Fx timeout reserved is in a follow up PR. This PR was scoped down to "port timeout functionality to batch".

File: test/integration-override/handlers/positions/handlerBatch.test.js (hunk 5 above)
Comment (kleyow) [reply test/integration-override/handlers/positions/handlerBatch.test.js]: A case of find and replace. reverted.

**Review by vijayg10 — APPROVED** [review]

**Pull request merged**